crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -260,103 +190,84 @@ impl<DB: Database> BlockProvider for ShareableDatabase<DB> {
+                {
+                    if let Some((header, block_hash)) = read_sealed_header(&self.tx, block_number)?
+                    {
+                        if let Some(block_body) =
+                            self.tx.get::<tables::BlockBodyIndices>(block_number)?
                         {
-                            if let Some((header, block_hash)) =
-                                read_sealed_header(tx, block_number)?
-                            {
-                                if let Some(block_body) =
-                                    tx.get::<tables::BlockBodyIndices>(block_number)?
-                                {
-                                    // the index of the tx in the block is the offset:
-                                    // len([start..tx_id])
-                                    // SAFETY: `transaction_id` is always `>=` the block's first
-                                    // index
-                                    let index = transaction_id - block_body.first_tx_num();
-
-                                    let meta = TransactionMeta {
-                                        tx_hash,
-                                        index,
-                                        block_hash,
-                                        block_number,
-                                        base_fee: header.base_fee_per_gas,
-                                    };
-
-                                    return Ok(Some((transaction.into(), meta)))
-                                }
-                            }
+                            // the index of the tx in the block is the offset:
+                            // len([start..tx_id])
+                            // SAFETY: `transaction_id` is always `>=` the block's first
+                            // index
+                            let index = transaction_id - block_body.first_tx_num();
+
+                            let meta = TransactionMeta {
+                                tx_hash,
+                                index,
+                                block_hash,
+                                block_number,
+                                base_fee: header.base_fee_per_gas,
+                            };
+
+                            return Ok(Some((transaction.into(), meta)))
                         }
                     }
                 }
+            }
+        }
 
-                Ok(None)
-            })?
-            .map_err(Into::into)
+        Ok(None)
     }
 
     fn transaction_block(&self, id: TxNumber) -> Result<Option<BlockNumber>> {
-        self.db
-            .view(|tx| {
-                let mut cursor = tx.cursor_read::<tables::TransactionBlock>()?;
-                cursor.seek(id).map(|b| b.map(|(_, bn)| bn))
-            })?
-            .map_err(Into::into)
+        let mut cursor = self.tx.cursor_read::<tables::TransactionBlock>()?;
```

Comment:
Would be rly nice if we could also pool cursors

Reply:
wdym by pooling exactly?

read txs are essentially read locks I think.

also the mdbx transaction can not be send across threads because it is not 'static